firedrake/dmplex.pyx

```diff
@@ -1014,3 +964,784 @@ def plex_renumbering(PETSc.DM plex, np.ndarray[PetscInt, ndim=1, mode="c"] reord
+               the direction of traversal.
+    :arg cell_ranks: MPI rank of the owner of each (visible) non-owned cell,
+                     or -1 for (locally) owned cell.
+    :arg orientations: Facet orientations relative to the plex.
+                       -1: orientation not set
+                        0: orientation is same as in (local) plex
+                       +1: orientation is the opposite of that in plex
+                       THIS ARRAY IS UPDATED BY THIS FUNCTION.
+
+    Returns the plex number of the facet visited last, or -1 if
+    'first_facet' is part of a closed loop.
+
+    Facets not incident to an owned cell are ignored.
+
+    Facet orientations are aligned to match the orientation, which was
+    assigned to orientations[first_facet - fStart].
+    """
+    cdef:
+        PetscInt fStart, fEnd
+        PetscInt from_facet = first_facet
+        PetscInt to_facet
+        PetscInt c = cell
+        np.int8_t plex_orientation
+
+        PetscInt local_from, local_to
+
+        PetscInt cone_size, support_size
+        PetscInt *cone = NULL
+        PetscInt *cone_orient = NULL
+        PetscInt support[2]
+        PetscInt i, ncells_adj
+
+    CHKERR(DMPlexGetHeightStratum(plex.dm, 1, &fStart, &fEnd))
+
+    # Retrieve orientation of first facet
+    plex_orientation = orientations[first_facet - fStart]
+
+    while True:
+        CHKERR(DMPlexGetConeSize(plex.dm, c, &cone_size))
+        assert cone_size == 4
+
+        CHKERR(DMPlexGetCone(plex.dm, c, &cone))
+        local_from = 0
+        while cone[local_from] != from_facet and local_from < cone_size:
+            local_from += 1
+        assert local_from < cone_size
+
+        local_to = (local_from + 2) % 4
+        to_facet = cone[local_to]
+
+        CHKERR(DMPlexGetConeOrientation(plex.dm, c, &cone_orient))
+        plex_orientation ^= (cone_orient[local_from] < 0) ^ True ^ (cone_orient[local_to] < 0)
+
+        # Store orientation of next facet
+        orientations[to_facet - fStart] = plex_orientation
+
+        if to_facet == first_facet:
+            # Closed loop
+            return -1
+
+        plex_get_restricted_support(plex, cell_ranks, to_facet, &support_size, support)
+
+        ncells_adj = 0
+        for i in range(support_size):
+            if support[i] != c:
+                ncells_adj += 1
+
+        if ncells_adj == 0:
+            # Reached boundary of local domain
+            return to_facet
+        elif ncells_adj == 1:
+            # Continue with next cell
+            for i in range(support_size):
+                if support[i] != c:
+                    from_facet = to_facet
+                    c = support[i]
+                    break
+        else:
+            assert ncells_adj > 1
+            raise RuntimeError("Facet belongs to more than two quadrilaterals!")
+
+    # We must not reach this point here.
+    raise RuntimeError("This should never happen!")
```

Comment:
OK, so this function takes a start facet, and a direction (which of the, possibly two, cells it's adjacent to) and propagates that facet's orientation as far as possible by orienting the "opposite" facet in the cell then moving to the next cell.